Added implimentation for TSK_FS_META and tsk_fs_file_read

AbdulRhmanAlfaifi opened the pull request:
Added the following :
* tsk_fs_meta.rs: A wrapper for TSK_FS_META.
* read file functions (read_exact,read_iter and read_to).
* TskFsFileIterator struct. An iterator over TskFsFile content bytes.

AbdulRhmanAlfaifi committed:
Added implimentation for TSK_FS_META and tsk_fs_file_read

Added the following :
* tsk_fs_meta.rs: A wrapper for TSK_FS_META.
* read file functions (read_exact,read_iter and read_to).
* TskFsFileIterator struct. An iterator over TskFsFile content bytes.

forensicmatt commented on src/tsk_fs_file.rs:
```diff
@@ -134,6 +136,54 @@ impl<'fs> TskFsFile<'fs> {
+    }
+
+    /// Reads the exact buffer size from the file
+    pub fn read_exact(&self,offset:i64, buff: &mut [u8]) -> Result<u64,TskError>{
```
Instead of read functionality being methods of `TskFsFile`, we will want to instead impl Read and Seek trait for `TskFsFile` or create a new struct like `TskFsFileHandle` that implements the Read/Seek trait. If Read and Seek trait is implemented you can pass the struct to any parser that parses based off of Read/Seek. For example, `TskFsAttr` implements Read/Seek, so you can use a lib like `mft` to do something like:

```rust
use mft::MftParser;

...
let mft_fh = tsk_fs.file_open_meta(0)
	.expect("Could not open $MFT");

let tsk_attr = mft_fh.get_attr()
	.expect("Unable to get default attribute.");

let mut parser = MftParser::from_read_seek(
  tsk_attr, 
  None
).expect("Error getting parser.");

let number_of_entries = parser.get_entry_count();
println!("number_of_entries: {}", number_of_entries);

for i in 0..number_of_entries {
	let entry = parser.get_entry(i as u64);
	...
}
```

This would also allow you to read a tsk file handle via normal means and write into another file like you would normally and not need something like `TskFsFileIterator`. The other thing we have to consider and the reason why we will most likely have to create a new struct like `TskFsFileHandle` is because we need the option to specify which attribute/data stream we would want to read from. For example, in the case of NTFS, you can have multiple data streams we would want to read from. A good example is $UsnJrnl, where you have Alternative Data Streams like $UsnJrnl:$Max and $UsnJrnl:$J. Thus, there has to be a mechanic in which you can specify what and how you want to read. Another item to account for is the [read flag](http://sleuthkit.org/sleuthkit/docs/api-docs/4.10.1//tsk__fs_8h.html#a9d4fee71e466b85026235e916ec2bea8) . Because of this, it is a much wiser choice to use the [tsk_fs_file_read_type](http://sleuthkit.org/sleuthkit/docs/api-docs/4.10.1//group__fslib.html#ga603ea8f80f1ffbf6be888aeaa508fc81) function when reading from a `TskFile` as it give you much more control on what and how to read. For that reason, I think it is better to create a new struct `TskFsFileHandle` that stores what and how reads will occur. `TskFsFile` should have methods therefor to get the `TskFsFileHandle` file from the `TskFsFile` and can either be from defaults or specified attributes and flags.

AbdulRhmanAlfaifi replied:
Ok, I like this better. I added a new struct called `TskFsFileHandler` which implements `Read` and `Seek`. The struct accepts `TskFsFile`, `TskFsAttrType` and `TskFsReadFlag`. `TskFsReadFlag ` is an enum and it is specific for the reading so it could be add to the file `tsk_fs_file_handler.rs`. However, the TskFsAttrType enum is the same types you already implemented on the TskFsAttr::type_name(). I think a better way of implementing it is:
* Add `TskFsAttrType` to `tsk_fs_file.rs`.
* Implement `Display` trait for `TskFsAttrType`
* change `type_name()` to use Display implementation.

Let me know if you are ok with these changes.

forensicmatt commented on src/tsk_fs_file.rs:
```diff
@@ -134,6 +136,54 @@ impl<'fs> TskFsFile<'fs> {
+    }
+
+    /// Reads the file in chuncks to the specified path
+    pub fn read_to(&self,path:&str) -> Result<bool,TskError>{
```
I would rather have Read/Seek traits implemented for TskFsFile. Because of that, it would be better for a tool that uses this library to determine how it wants to read and write to a file. I don't think this should be on the library level.

AbdulRhmanAlfaifi replied:
What about making the `TskFsFileHandler` accept `TskFsAttrType` & `TskFsFileReadFlag` and implementing `Read` & `Seek` on `TskFsFile` which only accepts `TskFsFileReadFlag` and read the attribute `NTFS_DATA` ?

forensicmatt commented on src/tsk_fs_file.rs:
```diff
@@ -450,3 +500,38 @@ impl<'fs, 'f> Iterator for TskFsAttrIterator<'fs, 'f> {
+/// * offset: keep track of the offset to be used for the next read operation.
+/// * file_size: the file size in bytes.
+/// * buf_size: the chunck size in bytes.
+pub struct TskFsFileIterator<'fs>{
```
This type of iteration is not needed with Read/Seek implementation

AbdulRhmanAlfaifi replied:
This struct has been removed

forensicmatt commented on src/tsk_fs_meta.rs:
```diff
@@ -0,0 +1,59 @@
+
+
+/// Wrapper for TSK_FS_META
+pub struct TskFsMeta(*const tsk::TSK_FS_META);
```
This struct wrapping is great!

AbdulRhmanAlfaifi replied:
Nice !

forensicmatt reviewed (changes requested):
Thanks for the code contribution! I think I want to keep this more Read/Seek and library friendly so I would like to make some changes.

AbdulRhmanAlfaifi committed:
Added TskFsFileHandler which is used to handle file read operations

* Added `TskFsFileHandler` to handler file read (implements Read & Seek traits).
* Added a wrapper for TSK_FS_META.
* moved `get_addr`,`is_dir` and `is_unallocated` functions from `TskFsFile` to `TskFsMeta`.
* Moved `TSK_FS_ATTR` wrapper structs to new file `tsk_fs_attr.rs`.
* Added `TskFsAttrType` enum to `tsk_fs_attr.rs`, changed `type_name` to return `TskFsAttrType` and impliminted `Display` for it.

forensicmatt commented on src/tsk_fs_attr.rs:
```diff
@@ -0,0 +1,348 @@
+use std::fmt::{Display, Formatter, Result as FmtReasult};
+
+#[derive(Copy, Clone, Debug)]
+pub enum TskFsAttrType {
```
I don't think I like having a libtsk-rs version of this when the bindings auto generate these for TSK_FS_ATTR_TYPE_ENUM.

This would require our lib to always maintain this list. Granted, I doubt it changes often, but either way, its just one more thing to upkeep and can limit people using the lib. Here you have listed 23 types, but there are actually 31 currently (thus, you can see how this is an issue):

```
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NOT_FOUND: TSK_FS_ATTR_TYPE_ENUM = 0;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_DEFAULT: TSK_FS_ATTR_TYPE_ENUM = 1;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_SI: TSK_FS_ATTR_TYPE_ENUM = 16;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_ATTRLIST: TSK_FS_ATTR_TYPE_ENUM = 32;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_FNAME: TSK_FS_ATTR_TYPE_ENUM = 48;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_VVER: TSK_FS_ATTR_TYPE_ENUM = 64;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_OBJID: TSK_FS_ATTR_TYPE_ENUM = 64;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_SEC: TSK_FS_ATTR_TYPE_ENUM = 80;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_VNAME: TSK_FS_ATTR_TYPE_ENUM = 96;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_VINFO: TSK_FS_ATTR_TYPE_ENUM = 112;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_DATA: TSK_FS_ATTR_TYPE_ENUM = 128;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_IDXROOT: TSK_FS_ATTR_TYPE_ENUM = 144;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_IDXALLOC: TSK_FS_ATTR_TYPE_ENUM = 160;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_BITMAP: TSK_FS_ATTR_TYPE_ENUM = 176;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_SYMLNK: TSK_FS_ATTR_TYPE_ENUM = 192;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_REPARSE: TSK_FS_ATTR_TYPE_ENUM = 192;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_EAINFO: TSK_FS_ATTR_TYPE_ENUM = 208;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_EA: TSK_FS_ATTR_TYPE_ENUM = 224;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_PROP: TSK_FS_ATTR_TYPE_ENUM = 240;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_NTFS_LOG: TSK_FS_ATTR_TYPE_ENUM = 256;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_UNIX_INDIR: TSK_FS_ATTR_TYPE_ENUM = 4097;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_UNIX_EXTENT: TSK_FS_ATTR_TYPE_ENUM = 4098;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_HFS_DEFAULT: TSK_FS_ATTR_TYPE_ENUM = 1;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_HFS_DATA: TSK_FS_ATTR_TYPE_ENUM = 4352;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_HFS_RSRC: TSK_FS_ATTR_TYPE_ENUM = 4353;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_HFS_EXT_ATTR: TSK_FS_ATTR_TYPE_ENUM = 4354;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_HFS_COMP_REC: TSK_FS_ATTR_TYPE_ENUM = 4355;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_APFS_DATA: TSK_FS_ATTR_TYPE_ENUM = 4352;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_APFS_RSRC: TSK_FS_ATTR_TYPE_ENUM = 4353;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_APFS_EXT_ATTR: TSK_FS_ATTR_TYPE_ENUM = 4354;
pub const TSK_FS_ATTR_TYPE_ENUM_TSK_FS_ATTR_TYPE_APFS_COMP_REC: TSK_FS_ATTR_TYPE_ENUM = 4355;
```

I think I would rather just use TSK_FS_ATTR_TYPE_ENUM instead where needed.

AbdulRhmanAlfaifi replied:
I agree but I hated dealing with bunch of const. But I found a way to generate the enum as rust enum using bindgen function `rustified_enum`. I implemented then new changes

forensicmatt commented on src/tsk_fs_attr.rs:
```diff
@@ -0,0 +1,348 @@
+/// 
+/// `fs => Filesystem lifetime
+/// 'f => File lifetime
+pub struct TskFsAttr<'fs, 'f>{
```
I do like this struct moved to its own file. Good idea!

forensicmatt commented on src/tsk_fs_file_handler.rs:
```diff
@@ -0,0 +1,170 @@
+}
+
+#[derive(Copy, Clone)]
+pub struct TskFsFileHandler<'fs>{
```
Lets call this `TskFsFileHandle` as its more generic and represents a File Handle. `TskFsFileHandler` implies more management type operations.

I think it will be easier for `TskFsFileHandle` to store an owned TskFsAttr. That is because both `attr_to_read` and `id`, and `attr_size` are all stored in the `TskFsAttr`. This will make for a cleaner implementation.

This will also make it easier to get `TskFsFileHandle`s from attributes, including a file's default attribute which makes things easy.

There are pieces of this code that are NTFS specific and would not work with other file systems.

AbdulRhmanAlfaifi replied:
I agree and done 👍

forensicmatt commented on src/tsk_fs_file_handler.rs:
```diff
@@ -0,0 +1,170 @@
+/// TskFsFileHandler struct is used to read file data.
+impl<'fs> TskFsFileHandler<'fs> {
+    /// Create TskFsFileHandler from TskFsFile. This will default to read $DATA attribute.
+    pub fn new(tsk_fs_file: &'fs TskFsFile<'fs>) -> Result<Self,TskError> {
```
I think `new` will need to have a `TskFsFile`, `TskFsAttr`, and read flag passed in.

forensicmatt commented on src/tsk_fs_file_handler.rs:
```diff
@@ -0,0 +1,170 @@
+    /// Create TskFsFileHandler from TskFsFile. This will default to read $DATA attribute.
+    pub fn new(tsk_fs_file: &'fs TskFsFile<'fs>) -> Result<Self,TskError> {
+        let attr_size = (&tsk_fs_file.get_meta()?).size();
+        let mut attr_to_read: TskFsAttrType = TskFsAttrType::NTFS_DATA;
```
This is a good example of code that would only work for NTFS file systems. Thus, we can't add statements like this.

AbdulRhmanAlfaifi replied:
Fixed

forensicmatt commented on src/tsk_fs_file_handler.rs:
```diff
@@ -0,0 +1,170 @@
+            _offset: 0,
+            attr_size,
+            attr_to_read,
+            read_flag: TskFsFileReadFlag::NoID, // Ignore the id and read $DATA
```
We shouldn't assume how the data is read, instead, the implementer needs to specify how to read.

forensicmatt commented on src/tsk_fs_file_handler.rs:
```diff
@@ -0,0 +1,170 @@
+            attr_size,
+            attr_to_read,
+            read_flag: TskFsFileReadFlag::NoID, // Ignore the id and read $DATA
+            id: 1
```
A hard coded id will not work. There is no guarantee that the attribute to read has an id of 1.

forensicmatt commented on src/tsk_fs_file_handler.rs:
```diff
@@ -0,0 +1,170 @@
+    /// You can call `get_attr` function on TskFsFile which will return `TskFsAttr` then call `get_attr_iter` function
+    /// which will return `TskFsAttrIterator` struct. Then iterate over `TskFsAttrIterator` to read all available attributes.
+    /// Pass the TskFsFile struct along with the id of the attribute you want to read.
+    pub fn new_with_id(tsk_fs_file: &'fs TskFsFile<'fs>, id: u16) -> Result<Self, TskError> {
```
Because `new()` should have the `TskFsAttr` passed in, there is no need to have a function that requires an id. Less work!

forensicmatt commented on src/tsk_fs_file_handler.rs:
```diff
@@ -0,0 +1,170 @@
+    }
+}
+
+impl<'fs> Read for TskFsFileHandler<'fs> {
```
Implementing Read/Seek... this is the way :)

forensicmatt commented on src/tsk_fs_meta.rs:
```diff
@@ -5,7 +5,66 @@ use crate::{
+}
+
+#[derive(Debug)]
+pub enum TskFsMetaType {
```
Again, not sure if I want to create an enum for something that is auto generated with the bindings. If anything, maybe all we want to do is implement `to_string` for `TSK_FS_META_FLAG_ENUM`, `TSK_FS_META_TYPE_ENUM`, etc.

AbdulRhmanAlfaifi replied:
I used `rustified_enum` function from bindgen to create rust enums instead of consts and implemented that.

AbdulRhmanAlfaifi committed:
Changes to TskFsHandler

* Change `TskFsFileHandler` to `TskFsFileHandle`.
* Change `TskFsFileHandle::new()` to accept TskfsAttr and read flag.
* Fix a bug on ``TskFsFileHandle` `Read` trait where it will not detect end of data properly.
* Added `id` function on `TskFsAttr` which return the attribute id.
* Added a new test case to copy a file.

AbdulRhmanAlfaifi commented:
* Change `TskFsFileHandler` to `TskFsFileHandle`.
* Change `TskFsFileHandle::new()` to accept TskfsAttr and read flag.
* Fix a bug on ``TskFsFileHandle` `Read` trait where it will not detect end of data properly.
* Added `id` function on `TskFsAttr` which return the attribute id.
* Added a new test case to copy a file.

forensicmatt commented:
@AbdulRhmanAlfaifi this is great! Thank you for your contribution. I will still make a couple tweaks (like renaming tsk_fs_file_handler.rs to tsk_fs_file_handle.rs, and TskFsFile::get_file_handler to TskFsFile::get_file_handle) but you did some solid work! This lib is still very alpha, so just keep in mind some things might change.
